designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,544 @@
+1. Remove `--rulesdir`
+1. Create a `@eslint/config` package to help users merge configs
+
+### The `.eslint.js` File
+
+The `.eslint.js` file is a JavaScript file (there is no JSON or YAML equivalent) that exports a `config` object: 
+
+```js
+exports.config = {
+    files: "*.js",
+    ignore: "*.test.js",
+    globals: {},
+    settings: {},
+    processor: object,
+    parser: object,
+    parserOptions: {},
+    ruledefs: {},
+    rules: {}
+};
+```
+
+The following keys are new to the `.eslint.js` format:
+
+* `files` - **Required.** Determines the global pattern that this configuration applies to and also is used as the default pattern to search for when ESLint is executed on the command line.
+* `ignore` - Determines the files that should not be linted using ESLint. This can be used in place of the `.eslintignore` file. The files specified by this glob pattern are subtracted from the files specified in `files`.
+* `ruledefs` - Contains definitions for rules grouped by a specific name. This replaces the `plugins` key in `.eslintrc` files and the `--rulesdir` option.
+
+The following keys are specified the same as in `.eslintrc` files:
+
+* `globals`
+* `settings`
+* `rules`
+* `parserOptions`
+
+The following keys are specified differently than in `.eslintrc` files:
+
+* `parser` - an object in `.eslint.js` files (a string in `.eslintrc`)
+* `processor` - an object in `.eslint.js` files (a string in `.eslintrc`)
+
+Each of these keys used to require one or more strings specifying module(s) to load in `.eslintrc`. In `.eslint.js`, these are all objects, requiring users to manually specify the objects to use.
+
+The following keys are invalid in `.eslint.js`:
+
+* `env` - responsibility of the user
+* `extends` - responsibility of the user
+* `overrides` - responsibility of the user
+* `plugins` - replaced by `ruledefs`
+* `root` - always considered `true`
+
+Each of these keys represent different ways of augmenting how configuration is calculated and all of that responsibility now falls on the user.
+
+#### Referencing Plugin Rules
+
+The `plugins` key in `.eslintrc` was an array of strings indicating the plugins to load, allowing you to specify processors, rules, etc., by referencing the name of the plugin. It's no longer necessary to indicate the plugins to load because that is done directly in the `.eslint.js` file. For example, consider this `.eslintrc` file:
+
+```yaml
+plugins:
+  - react
+
+rules:
+  react/jsx-uses-react: error
+```
+
+This file tells ESLint to load `eslint-plugin-react` and then configure a rule from that plugin. The `react/` is automatically preprended to the rule by ESLint for easy reference.
+
+In `.eslint.js`, the same configuration is achieved using a `ruledefs` key:
+
+```js
+const reactPlugin = require("eslint-plugin-react");
+
+exports.config = {
+    ruledefs: {
+        react: reactPlugin.rules
+    },
+    rules: {
+        "react/jsx-uses-react": "error"
```

Comment:
If we are making a departure from the current config, wouldn't it be easier to change the format to a nested object:
```js
rules: {
  react: {
    "jsx-uses-react": "error"
  }
}
```
This makes is a lot more machine readable, plus would be easier to override specific plugin configuration.

Reply:
:-1: I don't like that for a couple of reasons:

1. Makes it harder to reuse existing shareable configs because you know have to change the format.
2. With core rules that don't have a namespace, would we end up with core rule names at the same level as the namespace? I'm not a fan of mixing the meaning of key locations.